### ISSUE TITLE
Camera can move inside 3D terrain

### ISSUE THREAD

**Issue by mattgrid** [maplibre/maplibre-gl-js#1542]
This is a follow-on ticket from #1241. It's possible to move the camera to _within_ terrain when it's extruded from the 2D map surface.

@zdila said:

> May be a different issue, but in my case not only closest tiles but all. Open https://labs.maptiler.com/samples/maplibre/terrain/#style=satellite&lat=48.57344071&lng=20.46078220&zoom=17.16&bearing=-127.95&pitch=76.00&3d=true then zoom a little bit and tiles starts to disappear. If you refresh the page, nothing is shown until you zoom out. MapLibre 2.3.0.

@prozessor13 replied:

> Walking with the camera into terrain is still an open issue :/

### Expected behaviour

When the camera "bumps up" against 3D terrain, it should not be possible to move the camera to within the 3D surface.

### Actual behaviour

When you move the camera towards 3D terrain, nothing stops you when you "hit" the surface. You can keep moving through the surface.

**Comment by prozessor13** [comment]
I think i will not have time to look onto this in near future. Also I think before start implementing this, all possible scenarios should be played through. What should happen if, for example:

* map pitch changed with mouse-interaction, and camera collide with terrain.
    * may changing pitch should stop?
    * or should camera go up?
* map bearing changed with mouse-interaction, and camera collide with terrain.
    * I think camera should go up?
    * should it also rotate downwards do look onto the same center?
* flyTo/easeTo collide with terrain
* ...

**Comment by 767160** [comment]
My company would be happy to hire someone to fix this issue

**Comment by wipfli** [comment]
@767160 thanks for sharing. You can ping me on slack and I can maybe put you in touch with people who are interested... https://osmus-slack.herokuapp.com/

**Comment by 767160** [comment]
@wipfli Sorry, I do not use slack. Anyone interested, please send me an email at 061767@protonmail.com

**Comment by HarelM** [comment]
Assigned L bounty.
Link to parent Bounty: https://github.com/maplibre/maplibre/issues/189
If you would like to work on this and you think the bounty is not enough please ping me or contact @767160

**Comment by prozessor13** [comment]
I already have a basic working prototype for this, may in near future i find the time to create a PR.

**Comment by HarelM** [comment]
I've increased the bounty as I think L bounty isn't enough.

**Comment by prozessor13** [comment]
This is the basic working code, based on old versions of transform.ts and painter.ts, but i think the patch should work in the current classes as well.

Regards. Max.

```
diff --git a/src/geo/transform.ts b/src/geo/transform.ts
index 466fafaf7..bedc18210 100644
--- a/src/geo/transform.ts
+++ b/src/geo/transform.ts
@@ -520,6 +520,14 @@ class Transform {
         return {lngLat, altitude: altitude + this.elevation};
     }
 
+    // check that camera is always over terrain
+    checkTerrainCollision(terrain: Terrain) {
+        const camera = this.getCameraPosition();
+        const altitude = this.getElevation(camera.lngLat, terrain) + 100;
+        if (camera.altitude < altitude)
+            this.pitch = Math.acos((altitude - this._elevation) * this._pixelPerMeter / this.cameraToCenterDistance) / Math.PI * 180;
+    }
+
     /**
      * This method works in combination with freezeElevation activated.
      * freezeElevtion is enabled during map-panning because during this the camera should sit in constant height.
diff --git a/src/render/painter.ts b/src/render/painter.ts
index a707f3b21..61cd282b6 100644
--- a/src/render/painter.ts
+++ b/src/render/painter.ts
@@ -415,10 +415,16 @@ class Painter {
             // this is disabled, because render-to-texture is rendering all layers from bottom to top.
             this.opaquePassCutoff = 0;
 
+            // check if projMatrix has changed
+            if (!mat4.equals(this.terrainFacilitator.matrix, this.transform.projMatrix)) {
+               this.terrainFacilitator.dirty = true;
+               mat4.copy(this.terrainFacilitator.matrix, this.transform.projMatrix);
+               this.transform.checkTerrainCollision(this.style.map.terrain);
+            }
+
             // update coords/depth-framebuffer on camera movement, or tile reloading
             const newTiles = this.style.map.terrain.sourceCache.tilesAfterTime(this.terrainFacilitator.renderTime);
-            if (this.terrainFacilitator.dirty || !mat4.equals(this.terrainFacilitator.matrix, this.transform.projMatrix) || newTiles.length) {
-                mat4.copy(this.terrainFacilitator.matrix, this.transform.projMatrix);
+            if (this.terrainFacilitator.dirty || newTiles.length) {
                 this.terrainFacilitator.renderTime = Date.now();
                 this.terrainFacilitator.dirty = false;
                 drawDepth(this, this.style.map.terrain);
```

**Comment by typebrook** [comment]
This approach works great when changing bearing!

But there are some issues:
1. When I pan map toward a slope, map rendering stop and I got the following errors:
```error
Uncaught Error: failed to invert matrix         transform.ts:937:22    
    _calcMatrices transform.ts:937
    set pitch transform.ts:186
    checkTerrainCollision transform.ts:527
    render painter.ts:404
    ...
```

2. When camera is facing the bottom of a slope, I try to increase the pitch with `Ctrl` and mouse. Then the map shaking very quickly since pitch is changing by `checkTerrainCollision`.

**Comment by prozessor13** [comment]
For 1) please try:

```
diff --git a/src/geo/transform.ts b/src/geo/transform.ts
index bedc18210..4132d5687 100644
--- a/src/geo/transform.ts
+++ b/src/geo/transform.ts
@@ -523,9 +523,11 @@ class Transform {
     // check that camera is always over terrain
     checkTerrainCollision(terrain: Terrain) {
         const camera = this.getCameraPosition();
-        const altitude = this.getElevation(camera.lngLat, terrain) + 100;
-        if (camera.altitude < altitude)
-            this.pitch = Math.acos((altitude - this._elevation) * this._pixelPerMeter / this.cameraToCenterDistance) / Math.PI * 180;
+        const altitude = this.getElevation(camera.lngLat, terrain) + 10;
+        if (camera.altitude < altitude) {
+            const pitch = Math.acos((altitude - this._elevation) * this._pixelPerMeter / this.cameraToCenterDistance) / Math.PI * 180;
+            if (!isNaN(pitch)) this.pitch = pitch;
+        }
     }
 
     /**
```

**Comment by typebrook** [comment]
Yes, this make sense.

Would you like to change constant 100 by zoom level when assign `altitude`?

**Comment by prozessor13** [comment]
Yes, may 100 is to much! Feel free to make a proposal/change. This is only a quick and not very well tested solution!
Next, i will not create a PR and tests and all that stuff, may @SnailBones or you is willed?

**Comment by typebrook** [comment]
> may @SnailBones or you is willed?

Sure, I am willing to work on this. Since I am new to frontend. This issue seems the perfect one to help me realize map rendering workflow.

But @SnailBones has already leave the comment before,let @HarelM decide the assignee.



@prozessor13 
BTW, is it possible to check terrain collision in [Transform._constrain()]?

I notice this naming may related to change camera when terrain collision happens. But it seems `Terrain` is necessary to decide camera altitude.

[Transform._constrain()]: https://github.com/maplibre/maplibre-gl-js/blob/67944511634488e7ff05fd9de6f9d3d455dc90ec/src/geo/transform.ts#L746-L826

**Comment by prozessor13** [comment]
> > may @SnailBones or you is willed?
> 
> Sure, I am willing to work on this. Since I am new to frontend. This issue seems the perfect one to help me realize map rendering workflow.
> 
> But @SnailBones has already leave the comment before,let @HarelM decide the assignee.
> 
> @prozessor13 BTW, is it possible to check terrain collision in [Transform._constrain()](https://github.com/maplibre/maplibre-gl-js/blob/67944511634488e7ff05fd9de6f9d3d455dc90ec/src/geo/transform.ts#L746-L826)?
> 
> I notice this naming may related to **constrain** camera when terrain collision happens. But it seems `Terrain` is necessary to decide camera altitude.

Yes, this is the dilemma, that the Transform instance does nothing know about terrain. To be honest, my implementation was really a quick one, i did not go into detail. So it is really only a proposal! Feel free to change the code!

**Comment by HarelM** [comment]
Well @SnailBones was the first one to request, so it's up to him to decide if he would like to push this forward or leave it to @typebrook.

**Comment by SnailBones** [comment]
> Well @SnailBones was the first one to request, so it's up to him to decide if he would like to push this forward or leave it to @typebrook.

All yours @typebrook! Feel free to tag me for review.

**Comment by typebrook** [comment]
Thank you @SnailBones, I'll make a PR with tests in these days.

**Comment by typebrook** [comment]
@SnailBones @prozessor13 @HarelM

Sorry for the late reply. A new draft PR is made.

Since I am still not confident about realizing map rendering flow, if you guys find my approach is too naive, feel free to correct me or change the assignee.

**Comment by chrneumann** [comment]
@HarelM could you please assign this to me?

**Comment by HarelM** [comment]
Sure!
Have you registered with the developer pool?

**Comment by chrneumann** [comment]
Yes, I've talked to Oliver.
Thanks!

**Comment by HarelM** [comment]
I'm reopening this bug as it looks like the fix did not fully solve it.
I have panned and moved the map in the [doc site](https://maplibre.org/maplibre-gl-js/docs/examples/3d-terrain/) and it seems like I can still reach a state where the camera is inside the terrain - I did not do anything special, just used zoom and pan.
![image](https://github.com/user-attachments/assets/52903ce7-aba1-46ee-860f-ef8e6c413207)

**Comment by chrneumann** [comment]
It think it just looks as it is inside the terrain, because I removed the buffer in the last steps. The camera sits right on top of the terrain, which I think leads to this visual glitches. I would suggest to readd the buffers, that looks better.

The crash should be the one in https://github.com/maplibre/maplibre-gl-js/issues/4235 and less related to this issue.

**Comment by chrneumann** [comment]
![2024-08-13-111731_760x407_scrot](https://github.com/user-attachments/assets/ad5786c3-1291-4948-becd-e5cc83f7f84c)

Another example. The camera is not inside the terrain, but directly on the surface. This can be prevented by adding the buffer.

**Comment by HarelM** [comment]
So let try and add the buffer back and see if we don't get these issues anymore.

**Comment by chrneumann** [comment]
I think you can't prevent visual glitches completely, but adding the buffer would prevent the more common ones. For example, if a terrain on the left side of your view is really close and steep you might have to prevent the camera sitting next to it. But what if on the right side, there is also a steep hill very close? Should the camera be prevented to be in such a valley? But it might be ok if the field of view is narrow (the camera "fits" into the valley). But maybe the user resizes the canvas (and such the camera doesn't fit any more). There are so many factors to consider. I think that would make the solution way to complicated. So yeah, the buffer should prevent most, but will not prevent all visual glitches if you don't want to restrict camera movement too much.

As said, the crashes are something different, they must not occur. I would really suggest to handle them in a less catastrophic way. They should not lead the map to stop rendering. As the transformCamera implementiation depends on other code not directly manipulating the map's transform, there still might be places where this does happen and possibly will happen again with new code. I think it will take some time to harden this, and until then, problems with this should not lead to crashes. Maybe some form of linting that detects direct manipulation of the transform would be nice. Or really setting the transform private and only allow manipulation through the methods.

**Comment by HarelM** [comment]
We are doing a big refactoring of the trasform as part of the globe branch, so I hope it will be better afterwards. 
Let's add the buffet and make sure that most common cases are handled and that it's harder to create a "bad" scene.

**Comment by chrneumann** [comment]
Done in https://github.com/maplibre/maplibre-gl-js/pull/4551
Looks visually fine and the crashes are logged to console instead of thrown. Could not find any other issues.

**Comment by xen0l** [comment]
Any updates on this issue?

**Comment by clement-igonet** [comment]
I put together a side-by-side page to test this: left is current main, right is a candidate fix; the buttons run scripted camera journeys into terrain and the HUD shows camera altitude vs the ground under it.

Here is the "pan into slope" journey, with the same drag gestures sent to both sides:

![pan into slope, main vs fix](https://raw.githubusercontent.com/clement-igonet/maplibre-gl-js/assets/1542-pan-into-slope-compare.gif)

On main, every drag toward the slope pulls the camera lower, and it ends a few meters above the hillside, stuck in a blurry unusable view.

With the fix, the same gestures keep a safe clearance and the view stays usable the whole way.

The root cause seems to be that `Terrain._getOverscaledTileIDFromLngLatZoom` builds tile IDs from the fractional zoom, so the tile cache lookup never matches and the camera collision guard always reads elevation 0.

Flooring the zoom there is enough: the guard then sees real ground, and the harsher journeys on the page ("walk into mountain", "pitch into terrain"), which leave main's camera hundreds of meters inside the rock, stay clean.

**Comment by HarelM** [comment]
The right side looks as bad if not worse than the left side. Starting from a camera elevation of ~1000 km and ending at 4k isn't a better experience. Did you look at the gif?

**Comment by clement-igonet** [comment]
I tried to optimize the gif in order to get a light size. I'll try differently.

**Comment by HarelM** [comment]
The problem is not with the gif, is with the behavior it shows...

**Comment by clement-igonet** [comment]
You're right, and thanks for looking closely.
The clearance the guard applied was up to 500 m and it compounded across drags, which is what threw the right side into that aerial view.
I reduced it to at most 50 m and re-recorded, one journey per clip, same inputs to both sides.
Zooming toward the valley across a ridge, both sides now end at the same viewpoint; main transits about 800 m below the surface on the way, the fix never goes under:

![zoom into slope](https://raw.githubusercontent.com/clement-igonet/maplibre-gl-js/assets/1542-zoom-into-slope-v2.gif)

Walking the center toward Innsbruck at high pitch, main ends stuck about 580 m inside the ridge, the fix crests it:

![walk into mountain](https://raw.githubusercontent.com/clement-igonet/maplibre-gl-js/assets/1542-walk-into-mountain-v2.gif)

Pitching up at the foot of the wall, main ends about 520 m inside the rock, the fix stays just above the face:

![pitch into terrain](https://raw.githubusercontent.com/clement-igonet/maplibre-gl-js/assets/1542-pitch-into-terrain-v2.gif)

Repeated pans into the massif send main underground during the drags; the fix climbs over the ridge instead, ending at a comparable zoom:

![pan into slope](https://raw.githubusercontent.com/clement-igonet/maplibre-gl-js/assets/1542-pan-into-slope-v2.gif)

**Comment by HarelM** [comment]
I'm tired of talking to an AI bot.
Please stop this. I replied to your post in slack which you have yet to respond.
This is now just wasting my time and preventing me from reviewing people's work who actually spend the time looking at what they write and do.

**Comment by clement-igonet** [comment]
Well, currently it's me, not a bot. Sorry, I spend time to reply before I see you comment: "The problem is not with the gif, is with the behavior it shows...".
The 4 detailed behaviors not clearer ?

**Comment by HarelM** [comment]
Let's focus on the other PRs you opened first. Make sure you review thier code before I do. Make sure you test them manually and only after that request a review. Make sure to respond in the PR to cyder review comments.
After all that is completed we can return to this issue.
